rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -32,261 +34,388 @@ trait HotStore[F[_], C, P, A, K] {
   def snapshot(): F[Snapshot[C, P, A, K]]
 }
 
-final case class Cache[C, P, A, K](
-    continuations: TrieMap[Seq[C], Seq[WaitingContinuation[P, K]]] =
-      TrieMap.empty[Seq[C], Seq[WaitingContinuation[P, K]]],
-    installedContinuations: TrieMap[Seq[C], WaitingContinuation[P, K]] =
-      TrieMap.empty[Seq[C], WaitingContinuation[P, K]],
-    data: TrieMap[C, Seq[Datum[A]]] = TrieMap.empty[C, Seq[Datum[A]]],
-    joins: TrieMap[C, Seq[Seq[C]]] = TrieMap.empty[C, Seq[Seq[C]]],
-    installedJoins: TrieMap[C, Seq[Seq[C]]] = TrieMap.empty[C, Seq[Seq[C]]]
+final case class HotStoreState[C, P, A, K](
+    continuations: Map[Seq[C], Seq[WaitingContinuation[P, K]]] =
+      Map.empty[Seq[C], Seq[WaitingContinuation[P, K]]],
+    installedContinuations: Map[Seq[C], WaitingContinuation[P, K]] =
+      Map.empty[Seq[C], WaitingContinuation[P, K]],
+    data: Map[C, Seq[Datum[A]]] = Map.empty[C, Seq[Datum[A]]],
+    joins: Map[C, Seq[Seq[C]]] = Map.empty[C, Seq[Seq[C]]],
+    installedJoins: Map[C, Seq[Seq[C]]] = Map.empty[C, Seq[Seq[C]]]
 ) {
   def snapshot(): Snapshot[C, P, A, K] =
     Snapshot(
       this.copy(
-        continuations = this.continuations.snapshot(),
-        installedContinuations = this.installedContinuations.snapshot(),
-        data = this.data.snapshot(),
-        joins = this.joins.snapshot(),
-        installedJoins = this.installedJoins.snapshot()
+        continuations = this.continuations,
+        installedContinuations = this.installedContinuations,
+        data = this.data,
+        joins = this.joins,
+        installedJoins = this.installedJoins
       )
     )
 }
 
+private final case class HistoryStoreCache[F[_], C, P, A, K](
+    continuations: Map[Seq[C], Deferred[F, Seq[WaitingContinuation[P, K]]]],
+    datums: Map[C, Deferred[F, Seq[Datum[A]]]],
+    joins: Map[C, Deferred[F, Seq[Seq[C]]]]
+)
+
 private class InMemHotStore[F[_]: Concurrent, C, P, A, K](
-    cacheRef: Ref[F, Cache[C, P, A, K]],
-    historyReader: HistoryReaderBase[F, C, P, A, K]
+    hotStoreState: Ref[F, HotStoreState[C, P, A, K]],
```

Comment:
Can you please say to me: what is the point of Ref?

Reply:
`Ref` is An asynchronous, concurrent mutable reference. There is an easy example here https://typelevel.org/cats-effect/docs/std/ref.

You can wrap any data structure into a `Ref`and use the operation of `Ref` like `get`, `set` , `getAndUpdate` and etc. to make sure the data structure mutation is totally safe and consistent. The [example](https://typelevel.org/cats-effect/docs/std/ref#concurrent-counter) above is a very good example says how to mutate a data wrapped in a `Ref` in multi threads.

Reply:
Its related to https://github.com/rchain/rchain/pull/3526#discussion_r752948893, you can check https://typelevel.org/cats-effect/docs/std/ref also.